### PR TITLE
MPR#7638: problem with Mingw64 + bytecode + threads + exceptions raised from C

### PR DESCRIPTION
**Pull request by xavierleroy**
In the Windows Mingw64 port, multithreaded programs compiled to bytecode can crash when raising an exception from C code.  See [MPR#7638](https://caml.inria.fr/mantis/view.php?id=7638) for a repro and for more details.

This is a known problem with setjmp/longjmp in Mingw64.  This PR works around the issue by using GCC's __builtin_setjmp/__builtin_longjmp functions instead.  

Documentation about these builtins comes from https://patchwork.ozlabs.org/patch/290057/ especially the mysterious type `intptr_t[5]` for the jump buffer.

### CODE REVIEW

**Comment by alainfrisch** [comment]
Could you include the previously failing test in the testsuite?  (I guess there is no such test already, otherwise AppVeyor would have told us.)

**Comment by xavierleroy** [comment]
The existing test `tests/lib-threads/socketbuf.ml` exercises this bug -- that's how I found it!  Inria's CI was  failing lately because of this.  However, it could be that the problem shows up only in Windows 10, because earlier Inria's CI was Windows 7 (or 8?) and didn't fail.  I have no idea what configurations (of OCaml and of Windows) AppVeyor exercises.

Now there's also the "crashme" repro from the MPR that could be added as a test, if that's what you meant.

**Review by damiendoligez — APPROVED** [review]
LGTM. It seems the bug didn't occur in 4.02.3 so I'm trying to bisect which commit started triggering it. More info in a few hours.

**Comment by xavierleroy** [comment]
Acording to Inria's CI, the "socketbuf" test reproduces the problem with high probability but not always, probably because it is sensitive to thread scheduling.  The "crashme" test from the Mantis report should be reproducible.  For this reason I'm adding it to the test suite.

**Comment by dra27** [comment]
@xavierleroy - AppVeyor exercises Server 2012 R2 (at least when I last checked!). I'm concerned that I've never seen this - I think the crashme.ml should definitely go in as a test, please (I can repro that everywhere).

Microsoft have gone all Apple-y with Windows 10 - I'm afraid "Windows 10" no longer describes the OS in enough detail - you need the full version number or update name (easiest way is ver from a command prompt). I'm not seeing errors in socketbuf.ml on Windows 10 Creators (10.0.15063) or Server 2016 Anniversary (10.0.14393).

**Comment by dra27** [comment]
Sorry - didn't notice your comment in the meantime!

**Comment by xavierleroy** [comment]
OK, I added the "crashme" regression test as tests/lib-threads/pr7638.ml.

Concerning the exact version of Windows 10 on Inria's CI machines, I don't know, but maybe @damiendoligez does.

File: testsuite/tests/lib-threads/pr7638.ml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+(* MPR#7638 repro case *)
+
+let crashme v =
+  match Sys.getenv v with
+  | exception Not_found -> print_string "OK\n"
+  | s -> print_string "Surprising but OK\n"
```
Comment (dra27) [inline testsuite/tests/lib-threads/pr7638.ml:6]: :smile:

**Comment by xavierleroy** [comment]
And by the way: we should try to get you @dra27 full access (incl. SSH and RDP access) to the Inria CI machines, if you don't have it already and it is at all possible.

**Comment by shindere** [comment]
Xavier Leroy (2017/09/28 09:29 +0000):
> OK, I added the "crashme" regression test as tests/lib-threads/pr7638.ml.
> 
> Concerning the exact version of Windows 10 on Inria's CI machines, I
> don't know, but maybe @damiendoligez does.

Microsoft Windows [Version 10.0.15063]

**Comment by dra27** [comment]
@damiendoligez - it's triggering for me with 4.02.3

**Comment by shindere** [comment]
Xavier Leroy (2017/09/28 09:31 +0000):
> And by the way: we should try to get you @dra27 full access (incl. SSH
> and RDP access) to the Inria CI machines, if you don't have it already
> and it is at all possible.

Should be possible, yeah.

@dra27 should upload ssh public key to ci.inria.fr

Then we'lljust have to grant him access to the machine.

Can take care of that if you @xavierleroy and @damiendoligez so wish

**Comment by xavierleroy** [comment]
Thanks all for the quick reviews and feedback.  We're all anxious to see the various CIs green, so I'm merging right now.

**Comment by dra27** [comment]
@shindere - thanks, I'll get that sorted.

@xavierleroy - actually, thinking further, socketbuf.ml is a test which I'm used to having fail occasionally once, but rarely twice in row (so my testsuite tests pass because of TESTSUITE_RETRIES)

**Comment by xavierleroy** [comment]
Cherry-pick: [4.06 c5d265f]

**Comment by shindere** [comment]
David Allsopp (2017/09/28 09:41 +0000):
> @shindere - thanks, I'll get that sorted.

Just come back to me by private email when you are set-up.

> @xavierleroy - actually, thinking further, socketbuf.ml is a test
> which I'm used to having fail occasionally once, but rarely twice in
> row (so my testsuite tests pass because of TESTSUITE_RETRIES)

Would it help to be able to set a retry policy on a per-test basis?

Or should we, in theory, never have to retry?

**Comment by dra27** [comment]
@shindere - do I understand correctly from the new configuration that one slave builds the 4 Windows ports simultaneously now?

**Comment by dra27** [comment]
@shindere - in the perfect world, we'd never retry of course... the main reason for the TESTSUITE_RETRIES is that it mitigates some instances of Cygwin's fork randomly failing!

**Comment by dra27** [comment]
@damiendoligez - just as added information for your bisect. There must be environmental considerations to this too - I see the issue with 4.00.1 (last release of the first branch with mingw64 support) and Xavier's patch fixes that too.

**Comment by shindere** [comment]
David Allsopp (2017/09/28 09:47 +0000):
> @shindere - do I understand correctly from the new configuration that
> one slave builds the 4 Windows ports simultaneously now?

No. There is one slave per Windows port so 6 slaves as far as Windows is
concerned.

David Allsopp (2017/09/28 09:48 +0000):
> @shindere - in the perfect world, we'd never retry of course... the
> main reason for the TESTSUITE_RETRIES is that it mitigates some
> instances of Cygwin's fork randomly failing!

Would we be able to have a safer fork on Cygwin, i.e. a fork that is
itself retries when it fails?

**Comment by damiendoligez** [comment]
My bisect just pointed me to the commit that introduced the `socketbuf.ml` test in the first place. If you see the issue on 4.00.1, I don't think there's any point to bisecting anymore.

**Comment by damiendoligez** [comment]
> Would we be able to have a safer fork on Cygwin, i.e. a fork that is
itself retries when it fails?

That wouldn't work because it will fail again. Failure is mostly dependent on the addresses where various things are loaded, and that doesn't change until you restart the process.

**Comment by shindere** [comment]
Damien Doligez (2017/09/29 04:35 -0700):
> > Would we be able to have a safer fork on Cygwin, i.e. a fork that is
> itself retries when it fails?
> 
> That wouldn't work because it will fail again. Failure is mostly
> dependent on the addresses where various things are loaded, and that
> doesn't change until you restart the process.

Ah I see. I also understand why the retries strategy works. What I don't
understand, though, is why @dra27 mentionned somewhere that when fork
starts to fail one has to log out or restart the machine. Or did I
perhaps misunderstand you @dra27?

**Pull request merged**